### PR TITLE
aws.iam-role - delete action

### PR DESCRIPTION
**Pull request by PratMis**
By default AWS blocks any iam role deletion if its attached to an instance profile.

### CODE REVIEW

File: c7n/resources/iam.py [hunk 1]
```diff
@@ -817,6 +817,39 @@ def process(self, resources):
...
+    def process(self, resources):
+        client = local_session(self.manager.session_factory).client('iam')
+        for r in resources:
+            try:
+                client.delete_role(RoleName=r['RoleName'])
+            except ClientError as e:
+                if e.response['Error']['Code'] == 'DeleteConflict':
+                    self.log.warning(
+                        "Cannot delete entity, must remove roles from instance profile first")
+                if e.response['Error']['Code'] == 'NoSuchEntity':
```
Comment (thisisshi) [inline c7n/resources/iam.py]: this should continue instead of raising

Comment (PratMis) [reply c7n/resources/iam.py]: Gotcha! Changed to continue

File: c7n/resources/iam.py [hunk 2]
```diff
@@ -817,6 +817,41 @@ def process(self, resources):
...
+    For example, if you want to automatically delete an unused IAM role.
+
+    :example:
+
+      .. code-block:: yaml
+
+        - name: iam-delete-unused-role
+          resource: iam-role
+          filters:
+            - type: unused
```
Comment (kapilt) [inline c7n/resources/iam.py]: please switch the example to using usage, ... unused isn't really a safe filter, for many usages as it only considers ec2, ecs, lambda usages, and there are many other usage modes. the usage filter will look at actual api usage of the role.

File: c7n/resources/iam.py (hunk 1 above)
Comment (kapilt) [reply c7n/resources/iam.py]: can skip log here

File: c7n/resources/iam.py [hunk 3]
```diff
@@ -817,6 +817,41 @@ def process(self, resources):
...
+    schema = type_schema('delete')
+    permissions = ('iam:DeleteRole',)
+
+    def process(self, resources):
+        client = local_session(self.manager.session_factory).client('iam')
+        for r in resources:
+            try:
+                client.delete_role(RoleName=r['RoleName'])
+            except ClientError as e:
+                if e.response['Error']['Code'] == 'DeleteConflict':
```
Comment (kapilt) [inline c7n/resources/iam.py]: this needs to raise after processing the remainder.

File: c7n/resources/iam.py (hunk 1 above)
Comment (PratMis) [reply c7n/resources/iam.py]: Skipped log when role iam is not there

File: c7n/resources/iam.py (hunk 2 above)
Comment (PratMis) [reply c7n/resources/iam.py]: Switched to type:usage

File: c7n/resources/iam.py (hunk 3 above)
Comment (PratMis) [reply c7n/resources/iam.py]: Added code to raise at the end

**Comment by PratMis** [comment]
Thanks @kapilt for you feedback. I have made the changes as suggested.

File: c7n/resources/iam.py [hunk 4]
```diff
@@ -817,6 +817,44 @@ def process(self, resources):
...
```
Comment (kapilt) [inline c7n/resources/iam.py]: use error class from client.exceptions.DeleteConflictException and NoSuchEntityException.. thats preferred over string matching on error codes.

Comment (PratMis) [reply c7n/resources/iam.py]: Should we log in case of conflict?

File: c7n/resources/iam.py [hunk 5]
```diff
@@ -817,6 +817,44 @@ def process(self, resources):
...
+                    self.log.warning(
+                        "Cannot delete entity, must remove roles from instance profile first")
+                    continue
+                if e.response['Error']['Code'] == 'NoSuchEntity':
+                    continue
+                raise
```
Comment (PratMis) [inline c7n/resources/iam.py]: @kapilt If the exception is DeleteConflictException or NoSuchEntityException, don't think it'll ever hit raise in this case. Do you think we should remove it?

File: c7n/resources/iam.py (hunk 4 above)
Comment (kapilt) [reply c7n/resources/iam.py]:
```
error = None
for r in resources:
   try:
       do_something
   except NotFound:
        continue
   except OtherErrorContinue as e:
        error = e
if error:
   raise error
```

Comment (kapilt) [reply c7n/resources/iam.py]: a long for delete conflict makes sense

Comment (PratMis) [reply c7n/resources/iam.py]: Ah, ok! Understood! Thanks! I over thought it! I thought if there are multiple delete conflicts, we might have to capture all and raise a list.

File: c7n/resources/iam.py [hunk 6]
```diff
@@ -817,6 +817,45 @@ def process(self, resources):
...
+    permissions = ('iam:DeleteRole',)
+
+    def process(self, resources):
+        client = local_session(self.manager.session_factory).client('iam')
+        error = None
+        for r in resources:
+            try:
+                client.delete_role(RoleName=r['RoleName'])
+            except client.exceptions.DeleteConflictException as e:
+                self.log.warning(
```
Comment (kapilt) [inline c7n/resources/iam.py]: Log the actual role arn

Comment (PratMis) [reply c7n/resources/iam.py]: Sounds good

File: tests/test_iam.py [hunk 7]
```diff
@@ -1426,3 +1427,35 @@ def test_set_policy_detached(self):
...
+            "Statement": [{
+                "Effect": "Allow",
+                "Principal": {
+                    "Service": "ec2.amazonaws.com"
+                },
+                "Action": "sts:AssumeRole"
+            }]
+        })
+        client = factory().client("iam")
+        client.create_role(
```
Comment (kapilt) [inline tests/test_iam.py]: if your creating the role, should also be able to set this as a functional test, with the @functional decorator

File: c7n/resources/iam.py [hunk 8]
```diff
@@ -817,6 +817,46 @@ def process(self, resources):
...
+                    "Role %s cannot be deleted, must remove role from instance profile first"
```
Comment (kapilt) [inline c7n/resources/iam.py]: can you do this as role:%s it will play better with the json logging pr if so.

File: tests/test_iam.py (hunk 7 above)
Comment (PratMis) [reply tests/test_iam.py]: Cool

File: c7n/resources/iam.py (hunk 8 above)
Comment (PratMis) [reply c7n/resources/iam.py]: Sounds good

**Comment by kapilt** [comment]
your going to need additional tests around the failure scenarios to bring the coverage up.

**Comment by PratMis** [comment]
Ah, ok! I'll try to add more tests like a test to check if the role got created or not

**Comment by kapilt** [comment]
this more looks like you added a bunch of random recordings (and some extra get roles) to try and get coverage up. those recording's don't correspond to the test intent at all, ie service role doesn't correspond to tag filter in policy so would never be actioned. it would be better to make a separate *explicit* error test.

**Comment by PratMis** [comment]
Thanks, good catch! It was my personal account. Also I have created a separate test for error.

**Comment by kapilt** [comment]
looks like there is one left re account refs

**Comment by PratMis** [comment]
Apologies! Fixed now

**Pull request merged**